### PR TITLE
Added parsing of floating point number arguments.

### PR DESCRIPTION
**Pull request by hrittich**
This change is [Reviewable](https://reviewable.io/reviews/jarro2783/cxxopts/63)

### CODE REVIEW

File: include/cxxopts.hpp [hunk 1]
```diff
@@ -600,6 +610,18 @@ namespace cxxopts
 
     inline
     void
+    parse_value(const std::string& text, float& value) {
+      stringstream_parser(text, value);
+    }
+
+    inline
+	void
+	parse_value(const std::string& text, double& value) {
```
Comment (jarro2783) [inline include/cxxopts.hpp]: Remove the extra indents here.

Comment (hrittich) [inline include/cxxopts.hpp]: Done.

**Comment by jarro2783** [comment]
I'm sure I left a comment that seems to have been swallowed up somehow. See #60, can you make `stringstream_parser` the default for all `<T>`?

**Comment by hrittich** [comment]
The code should now work with types that overload >> .

**Comment by patrikhuber** [comment]
@jarro2783 I seem to be unable to use _any_ floating point values as command-line arguments, i.e.

```
float value;
("f,float", "A float", cxxopts::value<float>(value))
``` 

does not compile with compile error

> Severity	Code	Description	Project	File	Line	Suppression State
Error	C2665	'cxxopts::values::parse_value': none of the 10 overloads could convert all the argument types	test	\cxxopts\include\cxxopts.hpp	671	

Is that supposed to be the case without this PR?

Actually, I want to use a vector<float> as command-line argument. I see from the documentation that vector<string> works out-of-the-box, but not float. Is that supposed to be the case, and will this PR fix that?

Thank you!

**Comment by hrittich** [comment]
@patrikhuber Please see #60.

**Comment by patrikhuber** [comment]
@hrittich I read #60 already. It doesn't specifically talk about whether float parsing (not vector<float> but simple float) should be working right now (without this PR).

I'm sorry if I missed something but the documentation isn't very clear and `int` parsing currently seems to work so it's odd to me why `float` does not.

**Comment by hrittich** [comment]
@patrikhuber cxxopts contains a specialized parser for integers. This allows, e.g., to specify numbers in hexadecimal form, when they start with "0x". This specialized number parser, however, does not parse floating point numbers, which is the reason for this pull request.

**Comment by patrikhuber** [comment]
@hrittich I see! That makes complete sense. Thank you very much for the explanation! :-)

**Comment by patrikhuber** [comment]
I'm not sure whether this is working just yet. How can I make a `vector<float>` option work? I couldn't find anything in the documentation about `vector<T>` as options.

I tried:
`--myoption 1.0 0.0 -1.0`
`--myoption=1.0 0.0 -1.0`
`--myoption=1.0,0.0,-1.0`
`--myoption="1.0 0.0 -1.0"`
`--myoption="1.0,0.0,-1.0"`

I'm not sure what else to try?

**Comment by hrittich** [comment]
The following should work. You need to repeat the option name.
`--myoption 1.0 --myoption 0.0 --myoption -1.0`

**Comment by patrikhuber** [comment]
Oh wow, that's horribly ugly and inconvenient. Not really useful for my use case in fact.
Is there a way to change that into a more useful syntax without repeating the option name?

**Comment by hrittich** [comment]
The reason why cxxopts allows to use vectors for argument storage is to allow users to specify an argument more than once and the program to have access to all those arguments. The idea is not to actually have the user input something like a "logical" vector.

I do not see a suitable syntax that is easy to implement for your use case. For example, using a comma as a delimiter might fail for strings. The same is true for using spaces as delimiters. Someone would have to implement some rather complex parser. It would be possible, however, much work.

I would recommend you to parse the option as a string and then split it, the way you like.

**Comment by jarro2783** [comment]
Thanks @hrittich. Yes everything you said is correct.

You could also use a vector to hold the last positional argument, and then every positional argument that didn't go into another option would be added onto the end of that one.

The reason I made it like that is because a lot of programs like to allow an option to be written multiple times, so this is how you differentiate zero or one allowed, or more than one allowed.

If you wanted to parse your own you could set your value to be of some type `T`, and then provide an `operator>>(std::istream&, T&)` function.

**Comment by patrikhuber** [comment]
Thanks for the hints to both of you!

I tried the approach with a custom type but I did not succeed:

```
struct Coefficients
{
    std::vector<float> coefficients;
};


std::istream& operator>>(std::istream& is, Coefficients& c)
{
    double test;
    is >> test;
    c.coefficients.push_back(test);
    return is;
};
```
Then adding an option: `("shape-coeffs", "Docs", cxxopts::value<Coefficients>(coefficients))`
But if I now start my program with ` --shape-coeffs 1.0 2.0`, the `operator>>` overload only gets called once, for the `1.0`, and the `2.0` seems to get swallowed.

**Comment by hrittich** [comment]
The operator is only called whenever it encounters the option "--shape-coeffs". Then, you have to make sure that you read all the arguments that you think belong to the option, meaning, you somehow need to know how many numbers you have to read.

**Comment by patrikhuber** [comment]
I guess I could read up to the next long option (if there is any), that would work. But I have no idea from the cxxopts code I've seen so far how to do that. I think I may just go back to boost program_options unfortunately :-\ Or is that easy to accomplish?

**Comment by jarro2783** [comment]
There is no way for the parser to know that `--shape-coeffs` can take multiple arguments. You would have to just write it down multiple times. The other option would be to define your own type with an `operator>>`, and have the user write it as something like `1.0,2.0`.

**Comment by patrikhuber** [comment]
Yea I tried using my own type and define `operator>>`, but as you say, that would only work if there is no space between the numbers (like in `1.0,2.0`). I do not think it's a good option to put this burden on the user to don't allow white-spaces there, many will put white-space there, even if you clearly indicate otherwise in the documentation (I guess you know how "general" users are, too ;-) ).
I went back to boost program_options now. If cxxopts ever has facilities to accomplish this (and I don't think it's a too far-fetched use-case to be honest), I would be looking forward very much to switching back to it!

**Pull request merged**